core/src/banking_simulation.rs

```diff
@@ -0,0 +1,926 @@
+                (self.non_vote_tx_count - self.last_non_vote_count) as f64 / duration;
+            let tpu_vote_tps =
+                (self.tpu_vote_tx_count - self.last_tpu_vote_tx_count) as f64 / duration;
+            let gossip_vote_tps =
+                (self.gossip_vote_tx_count - self.last_gossip_vote_tx_count) as f64 / duration;
+            info!(
+                "senders(non-,tpu-,gossip-vote): tps: {:.0} (={:.0}+{:.0}+{:.0}) over {:?} not-recved: ({}+{}+{})",
+                tps,
+                non_vote_tps,
+                tpu_vote_tps,
+                gossip_vote_tps,
+                log_interval,
+                self.non_vote_sender.len(),
+                self.tpu_vote_sender.len(),
+                self.gossip_vote_sender.len(),
+            );
+            self.last_log_duration = simulation_duration;
+            self.last_tx_count = current_tx_count;
+            (
+                self.last_non_vote_count,
+                self.last_tpu_vote_tx_count,
+                self.last_gossip_vote_tx_count,
+            ) = (
+                self.non_vote_tx_count,
+                self.tpu_vote_tx_count,
+                self.gossip_vote_count,
+            );
+        }
+    }
+
+    fn on_terminating(self) {
+        info!(
+            "terminating to send...: non_vote: {} ({}), tpu_vote: {} ({}), gossip_vote: {} ({})",
+            self.non_vote_count,
+            self.non_vote_tx_count,
+            self.tpu_vote_count,
+            self.tpu_vote_tx_count,
+            self.gossip_vote_count,
+            self.gossip_vote_tx_count,
+        );
+    }
+
+    fn format_as_timestamp(time: SystemTime) -> impl Display {
+        let time: chrono::DateTime<chrono::Utc> = time.into();
+        time.format("%Y-%m-%d %H:%M:%S.%f")
+    }
+}
+
+impl BankingSimulator {
+    pub fn new(banking_trace_events: BankingTraceEvents, first_simulated_slot: Slot) -> Self {
+        Self {
+            banking_trace_events,
+            first_simulated_slot,
+        }
+    }
+
+    pub fn parent_slot(&self) -> Option<Slot> {
+        self.banking_trace_events
+            .freeze_time_by_slot
+            .range(..self.first_simulated_slot)
+            .last()
+            .map(|(slot, _time)| slot)
+            .copied()
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn prepare_simulation(
+        parent_slot: Slot,
+        first_simulated_slot: Slot,
+        genesis_config: GenesisConfig,
+        bank_forks: &Arc<RwLock<BankForks>>,
+        blockstore: &Arc<Blockstore>,
+        block_production_method: BlockProductionMethod,
+    ) -> (
+        Pubkey,
+        TracedSender,
+        TracedSender,
+        TracedSender,
+        BankWithScheduler,
+        PohService,
+        Arc<RwLock<PohRecorder>>,
+        BankingStage,
+        BroadcastStage,
+        Arc<LeaderScheduleCache>,
+        Sender<Slot>,
+        Arc<BankingTracer>,
+        TracerThread,
+        Arc<AtomicBool>,
+    ) {
```

Comment:
reminds of our loved `replay_stage.rs`...? lol I'm testing @apfitzge 's stress threshold..

Reply:
🫠 this seems like a lot.

If we immediately spawn the sender thread, can we just return it from this fn?
similarly i think we're preparing a bunch of stuff internal to the 2 loop fns - could we move some of that here and return simper objects that we can just `run`?

Reply:
thanks for sharing refactoring hint.

okay, i did another intense code shuffling: 6882957cd3f6d48a48b3b6dd24a84e0af748109d hope you like this..... now all `allow(clippy)` gone.

also, i fixed this bug 97fd71e0e6dc13e5c2f4e24c14c5c57d90b78a0e

Reply:
and it had bug: 42162825b2c88459c076dc183b00b475c3e51fc4